kafka/src/main/java/io/cloudevents/kafka/PartitionKeyExtensionInterceptor.java

```diff
@@ -0,0 +1,45 @@
+package io.cloudevents.kafka;
+
+import io.cloudevents.CloudEvent;
+import org.apache.kafka.clients.producer.ProducerInterceptor;
+import org.apache.kafka.clients.producer.ProducerRecord;
+import org.apache.kafka.clients.producer.RecordMetadata;
+
+import java.util.Map;
+
+/**
+ * This {@link ProducerInterceptor} implements the partitioning extension,
+ * as described in the <a href="https://github.com/cloudevents/spec/blob/master/kafka-protocol-binding.md#31-key-mapping">CloudEvents Kafka specification</a>.
+ * <p>
+ * When using in your producer, it will pick the {@code partitionkey} extension from the event and will set it as record key.
+ * If the extension is missing, It won't replace the key from the original record.
+ */
+public class PartitionKeyExtensionInterceptor implements ProducerInterceptor<Object, CloudEvent> {
+
+    public static final String PARTITION_KEY_EXTENSION = "partitionkey";
+
+    @Override
+    public ProducerRecord<Object, CloudEvent> onSend(ProducerRecord<Object, CloudEvent> record) {
+        if (record.value() == null) {
+            return record;
+        }
+        Object partitionKey = record.value().getExtension(PARTITION_KEY_EXTENSION);
+        if (partitionKey == null) {
+            return record;
+        }
+
+        return new ProducerRecord<>(record.topic(), record.partition(), record.timestamp(), partitionKey, record.value(), record.headers());
```

Comment:
the second field should be calculated based upon 1) the number of partitions, 2) the extension 3) the input record

Reply:
I don't think that's what the spec states: https://github.com/cloudevents/spec/blob/master/kafka-protocol-binding.md#31-key-mapping

I don't quite get what the implementation should look like to you, can you provide a sample code/pr of how you would implement it?

Reply:
In particular:

> Every implementation SHOULD provide a default "Key Mapper" implementation that maps the Partitioning partitionkey attribute value to the 'key' of the Kafka message as-is, if present.

It's named partitionkey, but in fact we're talking about the key of the message

Reply:
That's how this is implemented in other sdks too https://github.com/cloudevents/sdk-go/blob/master/protocol/kafka_sarama/v2/write_producer_message.go#L53